packages/cdktf/lib/terraform-resource-targets.ts

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright (c) HashiCorp, Inc.
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+import { TerraformResource } from "./terraform-resource";
+
+/**
+ *
+ */
+export class TerraformResourceTargets {
+  private _resourceTargetMap: Map<string, TerraformResource> = new Map<
+    string,
+    TerraformResource
+  >();
+
+  private prettyPrintEntries() {
+    return [...this._resourceTargetMap.entries()].reduce(
+      (accum, curr) => accum + "\n" + `${curr[0]}: ${curr[1].friendlyUniqueId}`,
+      ""
+    );
+  }
+
+  public addResourceTarget(resource: TerraformResource, target: string): void {
+    if (this._resourceTargetMap.has(target)) {
+      throw new Error(
+        `Target "${target}" has already been set at ${
+          this._resourceTargetMap.get(target)?.friendlyUniqueId
+        }`
+      );
+    }
+    this._resourceTargetMap.set(target, resource);
+  }
+
+  public getResourceByTarget(target: string): TerraformResource {
+    const result = this._resourceTargetMap.get(target);
+    if (!result) {
+      throw new Error(`Target "${target}" has not been set:
+      
+            Current Target Entries:\n ${this.prettyPrintEntries()}
```

Comment:
I'd change this a bit to something that says these are the currently available options. Also not sure but can this run into a race condition depending on the order of constructs?

Reply:
I think we could solve this by accessing it and throwing within a [LazyToken](https://github.com/hashicorp/terraform-cdk/blob/main/packages/cdktf/lib/tokens/lazy.ts#L110).

Reply:
A previous version had the possible bug that you mentioned (it failed if the target was not defined before the move).
However, in the current version we deferred that to the `toTerraform` method, which is only called once the construct tree is complete.

Reply:
Ah nice, that's great!